### ISSUE TITLE
revise the Food model

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#183]
8/18/20 design meeting: @amanda-phet @kathy-phet @ariel-phet @oliver-phet MH @pixelzoom 

Change requests for the food model:

* Swap wolves and food -- apply wolves at 4:00, food at 8:00
* When food is applied, do tough food first, then limited food
* No change to the tough food model (though @amanda-phet may revise parameter value).
* Limited food will reduce the population to the carrying capacity, which will be randomly selected from a range

**Comment by pixelzoom** [comment]
@amanda-phet and I discussed the issue of recessive mutants via Zoom.  Recessive mutants are currently excluded from application of environmental factors. This can lead to incorrect conclusions, and odd behavior if recessive mutants are the sole remaining bunnies.  We decided that this behavior is incorrect.  Recessive mutants can be favored during mating, to make them appear in the phenotype more quickly. But they should not be favored during selection.  I will change this.

**Comment by pixelzoom** [comment]
Moving the recessive mutant change to https://github.com/phetsims/natural-selection/issues/184.

**Comment by pixelzoom** [comment]
@amanda-phet this is ready for review in master or [1.1.0-dev.9](https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.9/phet/natural-selection_en_phet.html).  You'll likely want to tweak some parameters, then we can have the rest of the team take a look.

Here are the relevant query parameters. `limitedFoodPopulationRange` is new, with the default that was suggested by @kathy-phet during today's design meeting. `toughFoodPercentToStarveRange` was formerly `toughFoodPercentToKill`.

```
// The random percentage of bunnies that will die of starvation when food is tough.
toughFoodPercentToStarveRange: new Range( 0.45, 0.6 )
 
// Multiplier for bunnies with short teeth when food is tough.
shortTeethMultiplier: 1.6

// Range for the number of bunnies that can be sustained on limited food (carrying capacity).
limitedFoodPopulationRange: new Range( 90, 110 )
```

Tough food is still ignored if the population is < 5.  That's not available via a query parameter, it's a constant in Food.js:

```js
const TOUGH_FOOD_MIN_LONG_TEETH = 5;
```

Please also review the [Food section of model.md](https://github.com/phetsims/natural-selection/blob/master/doc/model.md#food).

**Comment by amanda-phet** [comment]
I've been playing and I think the following percentages make the most sense for tough food: 

`toughFoodPercentToStarveRange=.35,.5&shortTeethMultiplier=2`

I am pretty happy with the current limitedFoodPopulationRange, but @oliver-phet mentioned that he didn't think it seemed "limited" when where were a ton of bunnies hopping around. So I'd also be ok with `limitedFoodPopulationRange=60,80`, but I wouldn't go any lower than that. 

@kathy-phet @oliver-phet @ariel-phet do you want to play with the values at all?

**Comment by kathy-phet** [comment]
I think keeping the statistics useful is pretty important, so I would keep limitedFoodPopulationRange centered around 100. I played with it, and it seemed like a good amount to me ... basically it doesn't let the bunnies take over the world which is both useful, and a sign that the food is limited. 

I'll take a look at the sim with the food percenToStarve ranges you've chosen.

**Comment by pixelzoom** [comment]
@amanda-phet would you like me to change the defaults to `toughFoodPercentToStarveRange=.35,.5&shortTeethMultiplier=2` ?

And just to check... This can result in 100% of short-toothed bunnies being starved (0.5 * 2 = 1).

**Comment by oliver-phet** [comment]
> I think keeping the statistics useful is pretty important, so I would keep limitedFoodPopulationRange centered around 100.

The current limitedFoodPopulationRange=90,110 keeps the population around 270-330 for 2/3 of each generation and wolves then tough food are applied to that higher range before limited food is applied. @kathy-phet are you referring to the Proportions view?

> I've been playing and I think the following percentages make the most sense for tough food:
> 
> `toughFoodPercentToStarveRange=.35,.5&shortTeethMultiplier=2`

I might tweak the upper bound of the percent to starve to something like .4 or .45 so that it isn't possible to kill off 100% of the short-teeth bunnies. With .35,.5 I have to build up the population for 3 generations to reliably get a recessive long teeth population.

**Comment by kathy-phet** [comment]
Oliver - I’m referring to the number of bunnies left at the end of the generation.

**Comment by ariel-phet** [comment]
My only potential question/issue is with the short tooth bunnies being killed off completely.

If I change nothing else (using the parameters above `?toughFoodPercentToStarveRange=.35,.5&shortTeethMultiplier=2`) and start with a short toothed bunny population and add no mutations or such, pretty much no matter when I switch to tough food the bunnies all die out regardless of how many bunnies or such.

The piece of this interaction I find confusing is that although the bunnies take a serious population hit on each generation, they are able to mate and produce a bunch of bunnies often - that seems odd. If they are truly always going to starve to death no matter what, it seems strange that they would also be able to produce a bunch of offspring. If that is the goal, that basically without the mutation these bunnies will always die off, perhaps they should not mate as readily in the presence of tough food? 

I am fine leaving it the way it is, but it does seem a bit odd. Perhaps going to 1.8 would alleviate this concern.

**Comment by pixelzoom** [comment]
@ariel-phet said:
> The piece of this interaction I find confusing is that although the bunnies take a serious population hit on each generation, they are able to mate and produce a bunch of bunnies often - that seems odd. If they are truly always going to starve to death no matter what, it seems strange that they would also be able to produce a bunch of offspring. If that is the goal, that basically without the mutation these bunnies will always die off, perhaps they should not mate as readily in the presence of tough food?

In reality, yes, a food shortage would affect reproduction.  It would also affect lifespan.  That is not the case in this sim. It's a simplification that we decided on from the beginning (and which is also present in the Java version).  And the model is not at all capable of accommodating a change of this magnitude. It's also explained clearly in model.md:

> **Life Expectancy**
> 
> In reality, environmental factors may affect life expectancy. That is not the case in this simulation. All bunnies have an identical life expectancy of 5 generations.  If a bunny makes it to 5 generations old, it dies of old age. Bunnies die at 12:00, just before reproduction occurs.
> 
> **Reproduction**
> 
> In reality, environmental factors may affect reproduction. That is not the case in this simulation. Reproduction is not affected by environmental factors, and bunnies reproduce according to the following model:

@ariel-phet said:
> I am fine leaving it the way it is, but it does seem a bit odd. Perhaps going to 1.8 would alleviate this concern.

Do you mean reducing `shortTeethMultiplier` to 1.8?  That would actually result in fewer short-toothed bunnies starving due to tough food.

**Comment by ariel-phet** [comment]
@pixelzoom yes, I think fewer bunnies starving due to tough food makes a bit more sense (to me). I feel like with abundant (but tough) food, some short tooth bunnies would make it, but be easily outpaced by long toothed bunnies better adapted.

But if the learning goal is that certain creatures just cannot survive in certain environment, the short toothed always dying out is fine. My instincts tell me it would feel a bit more "realistic" with them being able to eat, but just not particularly effectively.

**Comment by oliver-phet** [comment]
> I am aware that the upper bound could kill 100% of the long-tooth bunnies. When the mutation is recessive (for fur color or teeth) and preferred, it should absolutely take over the population because you have a bunch of ff or tt bunnies mating with each other, so it would be expected for the entire undesirable population to eventually die. But making it 1.8 instead is fine.
> 
> If we change the percent to starve to .4 or .45, it will be exactly like the wolves, and I think we wanted food to be a little different.

I believe wolves are 0.35, 0.4 (2.3 multiplier for fur), so limited food could be .4, .45 (1.8 multiplier for teeth) and the factors would be different. 

I'm not sure what other feedback is requested of me, here's the list @pixelzoom originally added:

> - Swap wolves and food -- apply wolves at 4:00, food at 8:00

Looks great to me and helps with the limited food issues
> - When food is applied, do tough food first, then limited food

Looks great to me and helps with the limited food issues
> - No change to the tough food model (though @amanda-phet may revise parameter value).

I like ?toughFoodPercentToStarveRange=.4,.45&shortTeethMultiplier=1.8. I hesitate to make the multiplier 2.
> - Limited food will reduce the population to the carrying capacity, which will be randomly selected from a range

I like the "look" of a smaller range, but @kathy-phet wanted the population to min to be 100. The additional "food" minimizes my concern, too. So I think 90, 110 is fine.

**Comment by amanda-phet** [comment]
Thanks for the suggestion @oliver-phet . I am really happy with the behavior using `toughFoodPercentToStarveRange=.4,.45&shortTeethMultiplier=2`. I like keeping the multiplier higher and just decreasing the max percent a bit. 

I don't know how much longer we can wait on @kathy-phet 's feedback, so I am happy to make the decision here, especially given Megan's feedback in #204 . I think if she is happy with the sim behavior (which should remain true with this tweak to the tough food model), then we don't have anything to worry about. 

So, to sum up, the food model should be
`toughFoodPercentToStarveRange=.4,.45&shortTeethMultiplier=2` and 
`limitedFoodPopulationRange=90,110`

**Comment by kathy-phet** [comment]
I am planning to look at it tonight.

**Comment by kathy-phet** [comment]
I think overall its behaving well. I am wondering about one condition - when we have a batch of bunnies with both favored traits (long teeth and brown fur, say) and the fact that the death rate under the two conditions is so high that we don't really see them thriving after they dominate. They don't die out, but they don't really grow much either: 
https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.9/phet/natural-selection_en_phet.html?toughFoodPercentToStarveRange=.4,.45&shortTeethMultiplier=2&limitedFoodPopulationRange=90,110&labMutations=FT&labPopulation=100FFTT&screens=2

It's because the percentToStarveRange is .4-.45+ the percentToEatRange is .35-.4. So if we have 100 bunnies to start, all dominant, it should finish with a range of 33-39 bunnies. (or 33-39% survive). With birth we get a x3 increase in the population, so we rise back up to just the same population or very slightly larger. The result is that limited food is basically never a factor here. 

As long as Amanda and Megan have OKed the behavior under this particular scenario as desired, I think its working well under the other scenarios I tested. I do see that if the percentToStarveRange or percentToEatRange is lowered, under other scenarios the bunnies take over the world even faster and that might not be desirable. 

Back to you @amanda-phet

**Comment by kathy-phet** [comment]
Never mind, I've now looked at https://github.com/phetsims/natural-selection/issues/204 and its discussion of exactly what I was a bit worried about above, and Megan gave the behavior a thumbs up! Excellent. 

Unassigning @amanda-phet . 

Go with what Amanda said: 
So, to sum up, the food model should be
toughFoodPercentToStarveRange=.4,.45&shortTeethMultiplier=2 and
limitedFoodPopulationRange=90,110

**Comment by pixelzoom** [comment]
Parameters for the food model have been updated. 

@amanda-phet please verify in master or [1.1.0-dev.15](https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.15/phet/natural-selection_en_phet.html).  Run with `?log` to verify the complete set of sim-specific query parameters in the console:

```
query parameters: {  
  "allelesVisible": true,  
  "introMutations": "",  
  "introPopulation": [    "1"  ],  
  "labMutations": "",  
  "labPopulation": [    "1"  ],  
  "secondsPerGeneration": 1,  
  "fastForwardScale": 4,  
  "maxGenerations": 1000,  
  "maxPopulation": 750,  
  "maxAge": 5,  
  "mutationPercentage": 0.14285714285714285,  
  "wolvesPercentToEatRange": {  "_min": 0.35,  "_max": 0.4  }, 
  "wolvesEnvironmentMultiplier": 2.3,  
  "toughFoodPercentToStarveRange": {  "_min": 0.4,  "_max": 0.45  },  
  "shortTeethMultiplier": 2,  
  "limitedFoodPopulationRange": {  "_min": 90,  "_max": 110  },  
  "shrubsRange": {  "_min": 10,  "_max": 75  },  
  "showOrigin": false,  
  "showHorizon": false,  
  "showTimes": false
}
```

**Comment by amanda-phet** [comment]
That all looks good. Thanks!
